### PR TITLE
Starting with recipes

### LINKED ISSUE #855: Provide Apache proxy working example
Lots of people want LDAP backends, and rely on a proxy design.
I wish we had a reference Apache config alongside the nginx example (in `nginx.md`) so we can get rid of the burden answering people in that setup.

### PR DESCRIPTION
**Pull request by dmp42**
- apache recipe, including group level fine-grain authorization (enforce -B for @stevvooe)
- streamlined nginx recipe a bit to match better with Apache
- recipes file

Validated and fully tested accompanying shell scripts.

Fixes #855 and starts the recipe effort.

cc @docker/distribution-maintainers

### CODE REVIEW

**Comment by dmp42** [comment]
cc @thaJeztah for the usual laser-doc-review

File: docs/apache.md [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+# Authenticating proxy with apache
+
+## Use-case
+
+People already relying on an apache proxy to authenticate their users to other services might want to leverage it and have Registry communications tunneled through the same pipeline.
+
+Usually, that includes enterprise setups using LDAP/AD on the backend and a SSO mechanism fronting their internal http portal.
+
+## Alternatives
```
Comment (RichardScothern) [inline docs/apache.md]: Make this a sub header?
### Alternatives

File: docs/apache.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+## Solution
+
+With the method presented here, you implement basic authentication for docker engines in a reverse proxy that sits in front of your registry.
+
+While we use a simple htpasswd file as an example, any other apache authentication backend should be fairly easy to implement once you are done with the exemple.
+
+We also implement push restriction (to a limited user group) for the sake of the exemple. Again, you should modify this to fit your mileage. 
+
+## Gotchas
```
Comment (RichardScothern) [inline docs/apache.md]: Also a sub header?

File: docs/apache.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+While this model gives you the ability to use whatever authentication backend you want through the secondary authentication mechanism implemented inside your proxy, it also requires that you move TLS termination from the Registry to the proxy itself.
+
+Furthermore, introducing an extra http layer in your communication pipeline will make it more complex to deploy, maintain, and debug, and will possibly create issues.
+
+## Setting things up
+
+Read again the requirements.
```
Comment (RichardScothern) [inline docs/apache.md]: Which requirements?  Could you link to them?

File: docs/apache.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+</IfModule>
+
+<IfModule unixd_module>
+    User daemon
+    Group daemon
+</IfModule>
+
+ServerAdmin you@example.com
+
+ErrorLog /proc/self/fd/2
```
Comment (RichardScothern) [inline docs/apache.md]: Maybe note this is Linux/Unix only?

File: docs/apache.md [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+  SSLEngine on
+  SSLCertificateFile /usr/local/apache2/conf/domain.crt
+  SSLCertificateKeyFile /usr/local/apache2/conf/domain.key
+
+  # Higher Strength SSL Ciphers
+  SSLProtocol all -SSLv2 -SSLv3 -TLSv1
+  SSLCipherSuite RC4-SHA:HIGH
+  SSLHonorCipherOrder on
+
+  Header always set "Docker-Distribution-Api-Version" "registry/2.0"
```
Comment (RichardScothern) [inline docs/apache.md]: Comment for this section?

File: docs/apache.md [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+  SSLEngine on
+  SSLCertificateFile /usr/local/apache2/conf/domain.crt
+  SSLCertificateKeyFile /usr/local/apache2/conf/domain.key
+
+  # Higher Strength SSL Ciphers
+  SSLProtocol all -SSLv2 -SSLv3 -TLSv1
+  SSLCipherSuite RC4-SHA:HIGH
+  SSLHonorCipherOrder on
+
```
Comment (RichardScothern) [inline docs/apache.md]: Comment for this section

File: docs/apache.md [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    - `pwd`/data:/var/lib/registry
+
+EOF
+```
+
+## Starting and stopping
+
+Now, start your stack:
+
+    docker-compose up -d`
```
Comment (RichardScothern) [inline docs/apache.md]: extraneous (or missing) `

**Comment by RichardScothern** [comment]
A few nits, but a tasty recipe.

File: docs/apache.md (hunk 4 above)
Comment (dmp42) [inline docs/apache.md]: Mmmm... I see your point.
Now, this is really meant to be used with a specific hub official image, so, I think we are ok with that for now (people running this kind of s on windows is still a long shot).

File: docs/apache.md (hunk 6 above)
Comment (dmp42) [inline docs/apache.md]: Calling @diogomonica and @NathanMcCauley to the rescue on this: can you guys validate my tls babbling here is remotely proper?

File: docs/apache.md (hunk 7 above)
Comment (dmp42) [inline docs/apache.md]: Fixed.

File: docs/apache.md (hunk 3 above)
Comment (dmp42) [inline docs/apache.md]: Fixed. Also in nginx.md

File: docs/apache.md (hunk 2 above)
Comment (dmp42) [inline docs/apache.md]: Agreed.

File: docs/apache.md (hunk 1 above)
Comment (dmp42) [inline docs/apache.md]: Agreed.

**Comment by dmp42** [comment]
@RichardScothern PTAL

Also, do you think I should block PATCH as well? Do you know why it's not failing currently?

File: docs/apache.md (hunk 6 above)
Comment (aaronlehmann) [inline docs/apache.md]: You definitely shouldn't use RC4.

**Comment by dmp42** [comment]
@aaronlehmann PTAL on the SSL gymnastics.

**Comment by aaronlehmann** [comment]
The SSL stuff looks good to me. I was going to recommend the same page you cited.

File: docs/apache.md (hunk 4 above)
Comment (RichardScothern) [inline docs/apache.md]: Yes, probably not a big deal, although BSD doesn't have /proc either

**Comment by RichardScothern** [comment]
LGTM

File: docs/apache.md (hunk 4 above)
Comment (stevvooe) [inline docs/apache.md]: Maybe, something like this: `ErrorLog | sh -c 'cat 1>&2'`.

Comment (dmp42) [inline docs/apache.md]: I can't figure out a way to tame Apache to do that (re: http://httpd.apache.org/docs/2.4/mod/core.html#errorlog )

**Comment by dmp42** [comment]
Updated nginx TLS settings as well.
Will do some TLS automated testing in a few to validate all that.

Added PATCH to the list of methods in the Apache config.
Couldn't figure out a way to have better Apache Log statements but then...

That should be good enough for a start at this. Going to chase Apache struggling users around and have them wagtest it.

File: docs/apache.md (hunk 4 above)
Comment (stevvooe) [inline docs/apache.md]: This example doesn't work:

```
ErrorLog "|/usr/local/bin/httpd_errors"
```

This:

```
ErrorLog "|sh -c 'cat 1>&2'"
```

Comment (dmp42) [inline docs/apache.md]: Neither work for me...

Comment (thaJeztah) [inline docs/apache.md]: I wonder if you should completely replace the default configuration file, or just add this file to the "extras" dir and include it in the main configuration file.

This is what the standard configuration file uses for logging;

https://github.com/docker-library/httpd/blob/502ecf54a9db85f556028036973a9bce72eae8e8/2.4/Dockerfile#L57-L58

Comment (dmp42) [inline docs/apache.md]: The default config file is no good IMO - eg: not specific enough, and not secure enough. But yeah, basically they are doing the same thing as far as logging is concerned.

**Pull request merged**